provider/pkg/provider/util.go

```diff
@@ -133,3 +134,99 @@ func getActiveClusterFromConfig(config *clientapi.Config, overrides resource.Pro
 
 	return activeCluster
 }
+
+// pruneMap recursively drops keys from the source map that don't have a matching key in the target map and returns
+// the result. This is useful as a preprocessing step for live resource state before comparing it to program inputs.
+func pruneMap(source, target map[string]interface{}) map[string]interface{} {
+	result := make(map[string]interface{})
+
+	for key, value := range source {
+		if targetValue, ok := target[key]; ok {
+			valueT := reflect.TypeOf(value)
+			targetValueT := reflect.TypeOf(targetValue)
+
+			if valueT == nil || targetValueT == nil {
+				result[key] = value
+				continue
+			}
+
+			if valueT.Kind() == reflect.Map {
+				if targetValueT.Kind() == reflect.Map {
+					nestedResult := pruneMap(value.(map[string]interface{}), targetValue.(map[string]interface{}))
+					if len(nestedResult) > 0 {
+						result[key] = nestedResult
+						continue
+					}
+				} else {
+					result[key] = value
+					continue
+				}
+			}
+
+			if reflect.TypeOf(value).Kind() == reflect.Slice {
+				if reflect.TypeOf(targetValue).Kind() == reflect.Slice {
+					nestedResult := pruneSlice(value.([]interface{}), targetValue.([]interface{}))
+					if len(nestedResult) > 0 {
+						result[key] = nestedResult
+						continue
+					}
+				} else {
+					result[key] = value
+					continue
+				}
+			}
+
+			result[key] = value
+		}
+	}
+
+	return result
+}
+
+// pruneSlice drops elements from the source slice that don't have a matching element in the target slice
+func pruneSlice(source, target []interface{}) []interface{} {
```

Comment:
I don't see a comparison `targetValue` == `value` in case they are not maps or slices. For a target of `["a", "b"]` and a source of `["a", "c"]` should the result be `["a"]` or `["a", "c"]`? Couldn't spot a relevant test either but may have overlooked it.

Reply:
The result should be `["a", "c"]`. https://github.com/pulumi/pulumi-kubernetes/pull/2436/files#diff-95c076ba0b29c8c994e970141f5fa2789d9754af0f47da135a1dd94041809e70R956-R964 is the relevant test for this.

Although now I'm wondering if it should just leave the source slice alone if the elements are simple values.

Reply:
> Although now I'm wondering if it should just leave the source slice alone if the elements are simple values.

I tested this out more and think the current logic is correct.